### PR TITLE
Fixes behaviour for incomplete lines when `recover_with_nulls` is enabled

### PR DESCRIPTION
**Pull request by elstehle**
## Description

Closes https://github.com/rapidsai/cudf/issues/14227. Adapts the behaviour of the JSON finite-state transducer (FST) when `recover_with_nulls` is `true` to be more strict and reject lines that contain incomplete JSON objects (aka records) or JSON arrays (aka lists).

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/json/nested_json_gpu.cu
Comment (vuule) [inline cpp/src/io/json/nested_json_gpu.cu:1]: could you leave a comment on the changes in logic in the FST (how does this translate into the intended behavior change)? I can glean some logic from the code, but it's possibly made up :D

Comment (elstehle) [reply cpp/src/io/json/nested_json_gpu.cu:1]: I've added comments to the code (`get_translation_table` and the `nl_tokens` lambda. 

Essentially, we just changed the behaviour when we're in `recover_from_error`, such that we fail if the JSON line is an incomplete JSON value. E.g.,
- when we see a newline while we are still parsing a `LIST` or a `STRUCT`, we emit an `ErrorBegin` token that will mark this incomplete JSON line as invalid.
- when we see a newline while we are still parsing a `string` or `field name` (e.g., `{"a":"123\n`) , we emit an `ErrorBegin` token that will mark this incomplete JSON line as invalid.

**Review by vuule — APPROVED** [review]
Thank you for the comments.
It's still challenging to understand case-to-case why `regular_tokens` and `recovering_tokens` might be different, but I think I got it.

**Review by ttnghia — APPROVED** [review]

**Pull request merged**